### PR TITLE
Add store, load and CZ operations for qubit and resonator

### PR DESCRIPTION
**Pull request by rreiner-hqs**

### CODE REVIEW

File: qoqo/src/operations/spin_boson_operations.rs [hunk 1]
```diff
@@ -86,3 +86,68 @@ pub struct JaynesCummings {
     mode: usize,
     theta: CalculatorFloat,
 }
+
+
```
Comment (kbarkhqs) [inline qoqo/src/operations/spin_boson_operations.rs:90]:
```suggestion
```

File: roqoqo/src/operations/spin_boson_operations.rs [hunk 2]
```diff
@@ -154,3 +154,138 @@ impl SupportedVersion for JaynesCummings {
         (1, 10, 0)
     }
 }
+
+/// Stores a single excitation from the involved qubit into the involved bosonic mode as follows
+/// |0⟩_B ⨂ (a |0⟩_Q + b |1⟩_Q) -> (a|0⟩_B + b |1⟩_B ) ⨂ |0⟩_Q
+///
+/// Note: not defined if the bosonic mode is in a state |n> with n != 0
```
Comment (kbarkhqs) [inline roqoqo/src/operations/spin_boson_operations.rs:161]: I would suggest changing the `>` to the symbol used above

File: roqoqo/src/operations/spin_boson_operations.rs [hunk 3]
```diff
@@ -154,3 +154,138 @@ impl SupportedVersion for JaynesCummings {
...
+impl SupportedVersion for SingleExcitationStore {
+    fn minimum_supported_roqoqo_version(&self) -> (u32, u32, u32) {
+        (1, 10, 0)
+    }
+}
+
+/// Loads a single excitation from a bosonic mode into a qubit as follows
+/// (c1 |0⟩_B + c2 |1⟩_B) ⨂ |0⟩_Q -> |0⟩_B ⨂ (c1 |0⟩_Q + c2 |1⟩_Q)
+///
+/// Note: if the initial qubit state is |1>_Q the operation is only defined if c2 = 0
```
Comment (kbarkhqs) [inline roqoqo/src/operations/spin_boson_operations.rs:207]: Same docstring issue here with `>`

File: roqoqo/src/operations/spin_boson_operations.rs [hunk 4]
```diff
@@ -154,3 +154,138 @@ impl SupportedVersion for JaynesCummings {
...
+
+impl ImplementedIn1point10 for SingleExcitationLoad {}
+
+impl SupportedVersion for SingleExcitationLoad {
+    fn minimum_supported_roqoqo_version(&self) -> (u32, u32, u32) {
+        (1, 10, 0)
+    }
+}
+
+/// Controlled-Z operation between a qubit and a bosonic mode.
```
Comment (kbarkhqs) [inline roqoqo/src/operations/spin_boson_operations.rs:250]: Could you elaborate a little more in the docstring?

File: roqoqo/tests/integration/operations/spin_boson_operations.rs [hunk 5]
```diff
@@ -55,11 +71,26 @@ fn clone(op: Operation) {
 )]
 #[test_case(
     Operation::from(LongitudinalCoupling::new(4, 0, 1.5.into())),
-   "LongitudinalCoupling(LongitudinalCoupling { qubit: 4, mode: 0, theta: Float(1.5) })"  
+    "LongitudinalCoupling(LongitudinalCoupling { qubit: 4, mode: 0, theta: Float(1.5) })"
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/spin_boson_operations.rs:74]:
```suggestion
   "LongitudinalCoupling(LongitudinalCoupling { qubit: 4, mode: 0, theta: Float(1.5) })"  
```

File: roqoqo/tests/integration/operations/spin_boson_operations.rs [hunk 6]
```diff
@@ -55,11 +71,26 @@ fn clone(op: Operation) {
 )]
 #[test_case(
     Operation::from(LongitudinalCoupling::new(4, 0, 1.5.into())),
-   "LongitudinalCoupling(LongitudinalCoupling { qubit: 4, mode: 0, theta: Float(1.5) })"  
+    "LongitudinalCoupling(LongitudinalCoupling { qubit: 4, mode: 0, theta: Float(1.5) })"  
 )]
 #[test_case(
     Operation::from(JaynesCummings::new(4, 0, 1.5.into())),
-   "JaynesCummings(JaynesCummings { qubit: 4, mode: 0, theta: Float(1.5) })"  
+    "JaynesCummings(JaynesCummings { qubit: 4, mode: 0, theta: Float(1.5) })"
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/spin_boson_operations.rs:78]:
```suggestion
   "JaynesCummings(JaynesCummings { qubit: 4, mode: 0, theta: Float(1.5) })"  
```

File: roqoqo/tests/integration/operations/spin_boson_operations.rs [hunk 7]
```diff
@@ -151,20 +227,29 @@ fn substitute_subsitutemodes(op: SingleModeOperation, op_test: SingleModeOperati
...
+)]
+#[test_case(
+    Operation::from(SingleExcitationStore::new(1, 0)),
+    "SingleExcitationStore"
+)]
+#[test_case(
+    Operation::from(CZQubitResonator::new(1, 0)),
+    "CZQubitResonator"
+)]
+fn operate_one_mode_one_spin(op: Operation, name: &str) {
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/spin_boson_operations.rs:252]: I would suggest renaming this test to `operate_tags_hqslang`

File: roqoqo/tests/integration/operations/spin_boson_operations.rs [hunk 8]
```diff
@@ -179,22 +264,41 @@ fn operate_one_mode_one_spin(op: Operation, op_param: Operation, name: &str) {
...
+#[test_case(
+    Operation::from(LongitudinalCoupling::new(1, 0, "test".into())),
+    "LongitudinalCoupling"
+)]
+#[test_case(
+    Operation::from(JaynesCummings::new(1, 0, "test".into())),
+    "JaynesCummings"
+)]
+fn is_parametrized(op_param: Operation, name: &str) {
     assert!(!op.is_parametrized());
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/spin_boson_operations.rs:282]: I believe a test with `is_parametrized_false` would be appropriate for when these operations aren't parametrised, and to test the new ones

File: roqoqo/tests/integration/operations/spin_boson_operations.rs [hunk 9]
```diff
@@ -179,22 +264,41 @@ fn operate_one_mode_one_spin(op: Operation, op_param: Operation, name: &str) {
...
     assert!(op_param.is_parametrized());
 }
 
 #[test_case(SingleModeOperation::from(QuantumRabi::new(0, 0, 1.0.into())))]
 #[test_case(SingleModeOperation::from(LongitudinalCoupling::new(0, 0, 1.0.into())))]
 #[test_case(SingleModeOperation::from(JaynesCummings::new(0, 0, 1.0.into())))]
+#[tedt_case(SingleModeOperation::from(SingleExcitationLoad::new(1, 0)))]
+#[tedt_case(SingleModeOperation::from(SingleExcitationStore::new(1, 0)))]
+#[tedt_case(SingleModeOperation::from(CZQubitResonator::new(1, 0)))]
```
[the remark this answers is not in the document]
Comment (kbarkhqs) [reply roqoqo/tests/integration/operations/spin_boson_operations.rs:291]: Check locally that the tests run with the feature, as the pipeline won't catch this

File: roqoqo/tests/integration/operations/spin_boson_operations.rs [hunk 10]
```diff
@@ -331,6 +435,114 @@ fn jaynes_cummings_serde() {
...
+            Token::Str("qubit"),
+            Token::U64(0),
+            Token::Str("mode"),
+            Token::U64(0),
+            Token::StructEnd,
+        ],
+    );
+}
+
+#[cfg(feature = "serialize")]
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/spin_boson_operations.rs:510]: These three added tests could arguably be turned into one using `test_case`, but I don't mind either way

File: qoqo/src/operations/spin_boson_operations.rs [hunk 11]
```diff
@@ -86,3 +86,68 @@ pub struct JaynesCummings {
...
+/// Loads a single excitation from a bosonic mode into a qubit as follows
+/// (c1 |0⟩_B + c2 |1⟩_B) ⨂ |0⟩_Q -> |0⟩_B ⨂ (c1 |0⟩_Q + c2 |1⟩_Q)
+///
+/// Note: if the initial qubit state is |1>_Q the operation is only defined if c2 = 0
```
Comment (kbarkhqs) [inline qoqo/src/operations/spin_boson_operations.rs:94]: Same here with `>` and elaboration

File: qoqo/src/operations/spin_boson_operations.rs [hunk 12]
```diff
@@ -86,3 +86,68 @@ pub struct JaynesCummings {
...
+    InvolveQubits,
+    JsonSchema
+)]
+pub struct SingleExcitationStore {
+    qubit: usize,
+    mode: usize,
+}
+
+/// Controlled-Z operation between a qubit and a bosonic mode.
+/// Args:
```
Comment (kbarkhqs) [inline qoqo/src/operations/spin_boson_operations.rs:137]:
```suggestion
///
/// Args:
```

File: qoqo/src/operations/spin_boson_operations.rs [hunk 13]
```diff
@@ -86,3 +86,68 @@ pub struct JaynesCummings {
...
+pub struct SingleExcitationLoad {
+    qubit: usize,
+    mode: usize,
+}
+
+/// Stores a single excitation from the involved qubit into the involved bosonic mode as follows
+/// |0⟩_B ⨂ (a |0⟩_Q + b |1⟩_Q) -> (a|0⟩_B + b |1⟩_B ) ⨂ |0⟩_Q
+///
+/// Note: not defined if the bosonic mode is in a state |n> with n != 0
+/// Args:
```
Comment (kbarkhqs) [inline qoqo/src/operations/spin_boson_operations.rs:118]:
```suggestion
///
/// Args:
```

File: qoqo/src/operations/spin_boson_operations.rs [hunk 14]
```diff
@@ -86,3 +86,68 @@ pub struct JaynesCummings {
...
+    OperateSingleQubit,
+    InvolveQubits,
+    JsonSchema
+)]
+pub struct SingleExcitationStore {
+    qubit: usize,
+    mode: usize,
+}
+
+/// Controlled-Z operation between a qubit and a bosonic mode.
```
Comment (kbarkhqs) [inline qoqo/src/operations/spin_boson_operations.rs:136]: Same elaboration as in roqoqo

File: qoqo/tests/integration/operations/spin_boson_operations.rs [hunk 15]
```diff
@@ -333,6 +475,42 @@ fn test_pyo3_hqslang(name: &'static str, input_operation: Operation) {
...
     "JaynesCummings")]
+#[test_case(
+    Operation::from(SingleExcitationLoad::new(1, 0)),
+    vec![
+        "Operation",
+        "GateOperation",
+        "ModeGateOperation",
+        "SingleModeGateOperation",
+        "SingleQubitGateOperation",
+        "SingleQubitGate",
```
Comment (kbarkhqs) [inline qoqo/tests/integration/operations/spin_boson_operations.rs:486]: Once this has been removed in the roqoqo source file, this will fail

**Review by kbarkhqs — CHANGES_REQUESTED** [review]
Overall it looks almost ready, just the tags to change and some documentation to expand

File: roqoqo/tests/integration/operations/spin_boson_operations.rs (hunk 10 above)
Comment (rreiner-hqs) [reply roqoqo/tests/integration/operations/spin_boson_operations.rs:510]: I've spent some hours trying to do that but the `Token::Struct` wants an entry with a `'static` lifetime, which would complicate the test needlessly. Do you know of an easy way to solve this?

File: roqoqo/tests/integration/operations/spin_boson_operations.rs (hunk 8 above)
Comment (rreiner-hqs) [reply roqoqo/tests/integration/operations/spin_boson_operations.rs:282]: Added

File: roqoqo/src/operations/spin_boson_operations.rs [hunk 16]
```diff
@@ -154,3 +154,138 @@ impl SupportedVersion for JaynesCummings {
...
+}
+
+#[allow(non_upper_case_globals)]
+const TAGS_SingleExcitationStore: &[&str; 7] = &[
+    "Operation",
+    "GateOperation",
+    "ModeGateOperation",
+    "SingleModeGateOperation",
+    "SingleQubitGateOperation",
+    "SingleQubitGate",
```
[the remark this answers is not in the document]
Comment (rreiner-hqs) [reply roqoqo/src/operations/spin_boson_operations.rs:192]: That makes sense, I've removed it

File: roqoqo/src/operations/spin_boson_operations.rs [hunk 17]
```diff
@@ -137,13 +135,12 @@ pub struct JaynesCummings {
...
 
 #[allow(non_upper_case_globals)]
-const TAGS_JaynesCummings: &[&str; 7] = &[
+const TAGS_JaynesCummings: &[&str; 6] = &[
     "Operation",
     "GateOperation",
     "ModeGateOperation",
     "SingleModeGateOperation",
     "SingleQubitGateOperation",
-    "SingleQubitGate",
```
Comment (kbarkhqs) [inline roqoqo/src/operations/spin_boson_operations.rs:146]: Good catch, I missed this last time

**Review by kbarkhqs — APPROVED** [review]

**Pull request merged**